### PR TITLE
U/pgee/dm 982

### PR DESCRIPTION
**Pull request by jdswinbank**

### CODE REVIEW

File: include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    Control _ctrl;
+    //algorithms::ScaledFlux::KeyTuple _fluxCorrectionKeys;
+    meas::base::FluxResultKey _fluxResultKey;
+    afw::table::Key<float> _radiusKey;
+    afw::table::Key<float> _radiusForRadiusKey;
+    meas::base::FlagHandler _flagHandler;
+    meas::base::SafeCentroidExtractor _centroidExtractor;
+};
+
+}}}} // namespace lsst::meas::extensions::shapeHSM
```
Comment (jdswinbank) [inline include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h]: Namespace name in comment is wrong.

File: include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        afw::table::SourceRecord & measRecord,
+        afw::image::Exposure<float> const & exposure
+    ) const;
+
+    virtual void fail(
+        afw::table::SourceRecord & measRecord,
+        meas::base::MeasurementError * error=NULL
+    ) const;
+
+
```
Comment (jdswinbank) [inline include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h]: Too much whitespace.

File: include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    void _applyForced(
+        afw::table::SourceRecord & source,
+        afw::image::Exposure<float> const & exposure,
+        afw::geom::Point2D const & center,
+        afw::table::SourceRecord const & reference,
+        afw::geom::AffineTransform const & refToMeas
+    ) const;
+    std::string const & _name;
+    Control _ctrl;
+    //algorithms::ScaledFlux::KeyTuple _fluxCorrectionKeys;
```
Comment (jdswinbank) [inline include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h]: Why is this commented?

(I guess the answer is that it relates to the commented "FIXME" lines in `KronFluxAlgorithm.cc`; we'll consider those lines when we get there, but this line definitely shouldn't be left commented with no explanation.)

File: include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        KronAperture const& aperture
+        ) const;
+
+    void _applyForced(
+        afw::table::SourceRecord & source,
+        afw::image::Exposure<float> const & exposure,
+        afw::geom::Point2D const & center,
+        afw::table::SourceRecord const & reference,
+        afw::geom::AffineTransform const & refToMeas
+    ) const;
```
Comment (jdswinbank) [inline include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h]: Not enough white space.

File: python/lsst/meas/extensions/photometryKron/kronLib.i [hunk 5]
```diff
@@ -37,15 +37,39 @@ Interface to Kron magnitudes
...
 #include "lsst/afw/image/Utils.h"
 #include "lsst/afw/geom/Point.h"
-#include "lsst/meas/algorithms.h"
-#include "lsst/meas/extensions/photometryKron.h"
+#include "lsst/afw/math.h"
+#include "lsst/afw/image.h"
+#include "lsst/afw/detection.h"
+#include "lsst/afw/table.h"
+#include "lsst/meas/base.h"
+#define PY_ARRAY_UNIQUE_SYMBOL LSST_MEAS_BASE_NUMPY_ARRAY_API
```
Comment (jdswinbank) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: I'm not familiar with the guts of NumPy, but my understanding is that this should be unique across all possible extension modules. This looks like it conflicts with `meas_base`. Would `LSST_MEAS_EXT_PHOTKRON_NUMPY_ARRAY_API` or similar be better?

File: python/lsst/meas/extensions/photometryKron/kronLib.i [hunk 6]
```diff
@@ -37,15 +37,39 @@ Interface to Kron magnitudes
...
+
 %include "lsst/p_lsstSwig.i"
 
-%include "lsst/base.h"
-%import "lsst/meas/algorithms/algorithmsLib.i"
+%import "lsst/afw/math/mathLib.i"
+%import "lsst/afw/image/imageLib.i"
+%import "lsst/afw/detection/detectionLib.i"
+%import "lsst/afw/table/tableLib.i"
+
```
Comment (jdswinbank) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: Too much whitespace.

File: src/KronFluxAlgorithm.cc [hunk 7]
```diff
@@ -16,10 +16,12 @@
 #include "lsst/afw/coord/Coord.h"
 #include "lsst/afw/geom/AffineTransform.h"
 #include "lsst/afw/geom/ellipses.h"
-#include "lsst/meas/algorithms/Photometry.h"
 #include "lsst/meas/algorithms/PSF.h"
+#include "lsst/meas/algorithms/Photometry.h"
```
Comment (jdswinbank) [inline src/KronFluxAlgorithm.cc]: Should be  alphabetical

File: src/KronFluxAlgorithm.cc [hunk 8]
```diff
@@ -476,29 +412,56 @@ getPsfFactor(CONST_PTR(afw::detection::Psf) psf,
...
+    _fluxResultKey(
+        meas::base::FluxResultKey::addFields(schema, name, "flux from Kron Flux algorithm")
+    ),
+    _radiusKey(schema.addField<float>(name + "_radius", "Kron radius (sqrt(a*b))")),
+    _radiusForRadiusKey(schema.addField<float>(name + "_radiusForRadius",
+                                      "Radius used to estimate <radius> (sqrt(a*b))")),
+    _centroidExtractor(schema, name, true)
+{
+    static boost::array<meas::base::FlagDefinition,N_FLAGS> const flagDefs = {{
+        {"flag", "general failure flag, set if anything went wrong"},
```
Comment (jdswinbank) [inline src/KronFluxAlgorithm.cc]: Maybe nice to give "general" a capital G to match the other flags.

File: src/KronFluxAlgorithm.cc [hunk 9]
```diff
@@ -525,40 +488,41 @@ void KronFlux::_applyAperture(
...
     //
     double const psfFactor = getPsfFactor(exposure.getPsf(), aperture.getCenter(),
-                                          R_K_psf*ctrl.nRadiusForFlux, ctrl.maxSincRadius);
-    source.set(_fluxCorrectionKeys.psfFactor, psfFactor);
-    source.set(_fluxCorrectionKeys.psfFactorFlag, false); // i.e. good
+                                          R_K_psf*_ctrl.nRadiusForFlux, _ctrl.maxSincRadius);
+
+    //  FIXME:  This need to be done after fluxCorrection is implemented in meas_base
+    //source.set(_fluxCorrectionKeys.psfFactor, psfFactor);
+    //source.set(_fluxCorrectionKeys.psfFactorFlag, false); // i.e. good
```
Comment (jdswinbank) [inline src/KronFluxAlgorithm.cc]: I don't like leaving commented, "FIXME" code in the source -- we can't rely on folks remembering or happening to notice it in future. If we know work has to be done in the future, it should be filed as an issue in JIRA and the comment should refer to that issue.

Comment (jdswinbank) [inline src/KronFluxAlgorithm.cc]: Also, as long as those lines are commented/not in use, there's no use for the call to `getPsfFactor()`.

Comment (PaulPrice) [inline src/KronFluxAlgorithm.cc]: The compiler would optimise it out.

Comment (RobertLuptonTheGood) [inline src/KronFluxAlgorithm.cc]: If you're commenting out related sets of lines (and there's a reason not to just remove them), using the C preprocessing is often a better option, especially as it enables you to say why you're doing it;

#define OLD_API 0 /* ... */

#if  OLD_API
...
#endif


        R

Comment (jdswinbank) [inline src/KronFluxAlgorithm.cc]: But it will issue a warning; if we can avoid warnings, we should, so that when we see one it actually means something.

File: ups/meas_extensions_photometryKron.table [hunk 10]
```diff
@@ -1,4 +1,5 @@
 setupRequired(meas_algorithms)
+setupRequired( meas_base)
```
Comment (jdswinbank) [inline ups/meas_extensions_photometryKron.table]: Redundant space.

File: tests/Kron.py [hunk 11]
```diff
@@ -389,6 +486,71 @@ def getTolFlux(self, a, b, kfac):
...
+                        radius = forced.get("extensions_photometryKron_KronFlux_radius")
+                        xc, yc = wcs.skyToPixel(source.getCoord()) - afwGeom.Extent2D(warped.getXY0())
+                        for r, ct in [(radius, ds9.BLUE), (radius*kfac, ds9.CYAN),]:
+                            shape.scale(r/shape.getDeterminantRadius())
+                            ds9.dot(shape, xc, yc, ctype=ct, frame=2)
+
+                    try:
+
+                        #self.assertClose(source.get("extensions_photometryKron_KronFlux_flux"), forced.get("extensions_photometryKron_KronFlux_flux"), rtol=2.0e-4, atol=None)
+                        #self.assertClose(source.get("extensions_photometryKron_KronFlux_radius"), scale*forced.get("extensions_photometryKron_KronFlux_radius"), rtol=None, atol=1.0e-12)
```
Comment (jdswinbank) [inline tests/Kron.py]: Why are these commented?

File: tests/Kron.py [hunk 12]
```diff
@@ -389,6 +486,71 @@ def getTolFlux(self, a, b, kfac):
...
+                if source.get("extensions_photometryKron_KronFlux_flag"):
+                    continue
+                angleList = [45, 90,]
+                scaleList = [1.0, 0.5]
+                offsetList = [(1.23, 4.56), (12.3, 45.6)]
+
+                for angle in angleList:
+                 for scale in scaleList:
+                  for offset in offsetList:
+#angle, scale, offset in itertools.product(angleList, scaleList, offsetList):
```
Comment (jdswinbank) [inline tests/Kron.py]: What is this doing here?

File: tests/Kron.py [hunk 13]
```diff
@@ -16,12 +16,14 @@
 import lsst.pex.logging as pexLogging
 import lsst.pex.policy as pexPolicy
 import lsst.afw.detection as afwDetection
+import lsst.afw.coord as afwCoord
```
Comment (jdswinbank) [inline tests/Kron.py]: Alphabetical.

File: tests/Kron.py [hunk 14]
```diff
@@ -37,14 +39,111 @@
 
 #-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
 
-class KronPhotometryTestCase(unittest.TestCase):
+def makeGalaxy(width, height, flux, a, b, theta, dx=0.0, dy=0.0, xy0=None, xcen=None, ycen=None):
```
Comment (jdswinbank) [inline tests/Kron.py]: It seems like we're adding a lot of duplicate code here for no obvious reason. This `makeGalaxy` seems to be very similar to code in `KronPhotometryTestCase.makeAndMeasure()`. You could probably make a convincing argument that refactoring the code, breaking it up into smaller functions, isn't a bad idea, but this duplication seems to be uncalled for.

File: tests/Kron.py [hunk 15]
```diff
@@ -130,36 +228,35 @@ def measureKron(self, objImg, xcen, ycen, nsigma, kfac, nIterForRadius):
...
-        msConfig.algorithms["flux.kron"].enforceMinimumRadius = False
+        msConfig = measBase.SingleFrameMeasurementConfig()
+        msConfig.algorithms.names = ["base_SdssCentroid", "base_SdssShape", "extensions_photometryKron_KronFlux"]
+        msConfig.slots.centroid = "base_SdssCentroid"
+        msConfig.slots.shape = "base_SdssShape"
+        msConfig.slots.apFlux = "extensions_photometryKron_KronFlux"
+        msConfig.slots.modelFlux = None
+        msConfig.slots.psfFlux = None
+        msConfig.slots.instFlux = None
+        #msConfig.algorithms.names.remove("correctfluxes")
```
Comment (jdswinbank) [inline tests/Kron.py]: Why is this here? (Probably because you want to add it back when you get to the FIXME above, but I think this isn't the right way to handle that).

File: tests/Kron.py [hunk 16]
```diff
@@ -259,7 +356,7 @@ def measureKronInPython(self, objImg, xcen, ycen, nsigma, kfac, nIterForRadius,
 
         return R_K, sumI, 0, False
 
-    def testEllipticalGaussian(self):
+    def xtestEllipticalGaussian(self):
```
Comment (jdswinbank) [inline tests/Kron.py]: I assume the 'x' is a cheap & cheerful way of skipping the test, but you'll need to explain properly why this test code should be included by not executed.

File: tests/Kron.py [hunk 17]
```diff
@@ -389,6 +486,71 @@ def getTolFlux(self, a, b, kfac):
...
+
+                    try:
+
+                        #self.assertClose(source.get("extensions_photometryKron_KronFlux_flux"), forced.get("extensions_photometryKron_KronFlux_flux"), rtol=2.0e-4, atol=None)
+                        #self.assertClose(source.get("extensions_photometryKron_KronFlux_radius"), scale*forced.get("extensions_photometryKron_KronFlux_radius"), rtol=None, atol=1.0e-12)
+                        self.assertEqual(source.get("extensions_photometryKron_KronFlux_flag"), forced.get("extensions_photometryKron_KronFlux_flag"))
+                    except:
+                        print ("Failed:", angle, scale, offset,)
+                             #  [(source.get(f), forced.get(f)) for f in
+                             #   ("flux.kron", "flux.kron.radius", "flux.kron.flags")])
```
Comment (jdswinbank) [inline tests/Kron.py]: Please don't include unused, commented code without good reason.

File: tests/Kron.py [hunk 18]
```diff
@@ -37,14 +39,111 @@
...
+
+def makeSourceMeasurementConfig(forced=False, nsigma=6.0, nIterForRadius=1, kfac=2.5):
+    """Construct a SourceMeasurementConfig with the requested parameters"""
+    if forced:
+        msConfig = measBase.ForcedMeasurementConfig()
+        msConfig.algorithms.names = ["base_TransformedCentroid", "base_SdssShape", "extensions_photometryKron_KronFlux"]
+        msConfig.slots.centroid = "base_TransformedCentroid"
+    else:
+        msConfig = measBase.SingleFrameMeasurementConfig()
+        msConfig.algorithms.names = ["base_SdssCentroid", "base_SdssShape", "extensions_photometryKron_KronFlux"]
```
Comment (jdswinbank) [inline tests/Kron.py]: Need to add `"base_SkyCoord"` here (NB not in line 97) or the test will fail when `display=True`; see discussion on JIRA.

File: tests/Kron.py (hunk 16 above)
Comment (PaulPrice) [inline tests/Kron.py]: The way to skip a test is to use the `@unittest.skip("skipping because…")` decorator.

File: tests/Kron.py (hunk 12 above)
Comment (PaulPrice) [inline tests/Kron.py]: The `product` has been split into separate `for` loops.  Note that the indentation is wrong, and splitting it up like this is unnecessary because the individual elements aren't used separately.

Comment (pgee2000) [inline tests/Kron.py]: I was sloppy to have left this in the unit test.  But the reason for it was
that the product notation didn't work with the python I was using, and I
wasn't sure why.  I split it up as a temporary measure, but forgot to come
back to it later and make a decision about how the code should really be.

File: include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h (hunk 3 above)
Comment (pgee2000) [inline include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h]: In commenting out lines, I was trying to save code which was previously used for flux correction, in hopes that whoever implemented that (I assumed me) would get some help from it.  But I agree that it should be annotated.  However, I got lots of comments on this code saying that I should never have commented code, even for a good reason.  I think some comment about what is missing is important, however, even if it is a back reference to the meas_algorithms code from which this came.

File: python/lsst/meas/extensions/photometryKron/kronLib.i (hunk 5 above)
Comment (pgee2000) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: I don't even know the answer to this.  It was cut and pasted from meas_base.  Is there a way to send this to jbosch from the github pull request?

Comment (TallJimbo) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: Yes, these need to be unique across all extension models.  @jdswinbank's suggestion is correct.  (And by the way, to get my attention here on GitHub, use @TallJimbo, but I was already watching this PR anyway).

File: src/KronFluxAlgorithm.cc (hunk 8 above)
Comment (pgee2000) [inline src/KronFluxAlgorithm.cc]: I actually think that Jim's example in PsfFlux.cc has lower case.  The upper case was from the older meas_algorithm doc strings, so I am taking lower case to be the rule and changing others to match "general".

File: tests/Kron.py (hunk 16 above)
Comment (pgee2000) [inline tests/Kron.py]: Just a failure to look closely before I commited.  This test takes a long time, and I was "temporarily" commenting it out.  Sloppy me.

File: include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h (hunk 3 above)
Comment (jdswinbank) [inline include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h]: I think that

> saying that I should never have commented code, even for a good reason

is not accurate: if there's a good reason, then of course you should do it. But, in general, a good reason is pretty hard to find.

Given that the whole history of the codebase is stored in git and that commented code like this adds potentially confusing noise for anybody who isn't trying to (re-)implement the feature in question, I think it should be gone (with a note in the JIRA ticket requesting the reimplementation which points out where it can be found). If you really, really think it's necessary and that the git history is inadequate, you should make sure the commented code is clearly annotated to say what it's for and why it's required.

File: tests/Kron.py (hunk 14 above)
Comment (pgee2000) [inline tests/Kron.py]: This is just the way the test was as I got it, except that the makeGalaxy

Comment (pgee2000) [inline tests/Kron.py]: Sorry, I meant to say that makeGalaxy was part of the original code, and I don't think that I even looked at it.  Glad to try to make it better if you think it is worth the effort, but please put it on a different ticket.

File: include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h (hunk 3 above)
Comment (pgee2000) [inline include/lsst/meas/extensions/photometryKron/KronFluxAlgorithm.h]: OK, that's perfectly reasonable.  I was trying to save time, not add
confusion.  With many of the code conversions I have done, the code cannot
be converted for one reason or another.  I will leave the old code in where
it seems useful, and add a descriptive comment where it is not.

File: python/lsst/meas/extensions/photometryKron/kronLib.i (hunk 5 above)
Comment (pgee2000) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: I'm not exactly sure why this line is here at all.  I copied it from
another .i file, and thought it was needed.  Please explain what it does.

I am also unclear about the problem with product.  It did not seem to work
when I reenabled testForced in Kron.py.  I worked around it, meaning to ask
you if it worked in previous versions of this tests.

Comment (TallJimbo) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: A line like this is needed any time you want to use one of `ndarray`s converters to/from `NumPy` arrays, including those for Eigen objects.  It looks like you're declaring at least one such converter below (the call to `%declareNumPyConverters(...)`.  My guess is that you don't actually need that here, since I don't think the Kron code ever measures a centroid, and hence you could remove this line and all the others that include ndarray or NumPy headers, and the `%init` block that initializes the NumPy library.

Comment (pgee2000) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: So if I did need these converters, is the right way to declare the define
and all the other stuff myself, or is there something I should be including
or importing from meas_base?

Comment (pgee2000) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: Do you have an answer about the itertools.product?  I didn't see anything
in the Python website indicating that it shouldn't work, but it seemed to
confuse the Python I was running.

Comment (TallJimbo) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: If you do need the converters, you would have to declare them all here, and repeat all the boilerplate.

I don't know what the question was about itertools.product; are you replying to the right thread on the PR?

Comment (PaulPrice) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: `itertools.product` should work with python 2.7.  Perhaps you were using an older version?

Comment (pgee2000) [inline python/lsst/meas/extensions/photometryKron/kronLib.i]: Not intentionally.  But I will check.

Thanks.

File: tests/Kron.py (hunk 14 above)
Comment (jdswinbank) [inline tests/Kron.py]: I think there's something more complex going on here, actually. To elaborate:

The `makeGalaxy` function was introduced by Paul in cc2925d several months ago:
```
$ git log --all --format="%h, %an, %ar: %s" -SmakeGalaxy
4ed8e0a, pgee, 3 weeks ago: DM-982 Convert kron algorithm to meas_base
fe56449, pgee, 2 weeks ago: Additional fixups:
cc2925d, Paul Price, 7 months ago: tests: add test for forced photometry under rotation, translation, scaling
```
Note that when Paul added made that change, he didn't just add the `makeGalaxy` function, but *moved* code from `KronPhotometryTestCase` into that function.

However, cc2925d is not part of the history of your branch:
```
$ git branch -r --contains cc2925d
  origin/HEAD -> origin/master
  origin/master
  origin/tickets/DM-1783
```
Instead, you seem to have introduced `makeGalaxy` yourself in 4ed8e0a. This is (as far as git is concerned) completely independent of Paul's code -- it looks as though you copied and pasted it, or something. However, you only copied the additions from Paul's commit (ie, the creation of the ``makeGalaxy`` function), not the removal of the code from elsewhere in that file. That means you've ended up with a whole bunch of duplication.

So what does that mean? First, you have to clean this up before you can merge it; it can't wait to go on a separate ticket. Secondly, it's pretty confusing for both you and Paul to have apparently spontaneously written exactly the same code; you need to clean up history so that that doesn't appear to have been the case.

My suggestion would be to rebase your changes on top of `master` before merging. As part of that rebase, you'll be adding Paul's cc2925d to the history of your branch. When you try to apply your own 4ed8e0a on top of it, you'll probably find that it conflicts. You'll need to resolve the conflict, effectively by deleting your copy of `makeGalaxy` and just using the one Paul wrote.

It's interesting to note that if you'd cleaned up your branch and rebased it against master before asking for a review, you'd already have had to fix this problem and I wouldn't have had to spend time questioning it and figuring out what happened. It's a nice example of the benefits of pre-review rebasing, as we were discussing just the other day!

Comment (pgee2000) [inline tests/Kron.py]: I guess that's right.  I looked at master to see what the original was,
when I should have looked at the origin of my branch.

OK, I will clean up before merging.  Sorry to put you through the trouble.

If you are saying that I should merge to master before asking for review, I
can do that.  That has not been part of the procedure in the past, but I'm
happy to do that.  It is rebasing against my own branch which I found not
very useful.

Comment (jdswinbank) [inline tests/Kron.py]: Thanks Perry!

One thing, though: I am *definitely not* saying you should merge to master before asking for review. I'm suggesting that you should at least consider *rebasing* your branch against the latest master. That makes it easier for the reviewer to see how your changes will integrate with the rest of the codebase, and it will make it much easier to merge the code after the review is complete. Plus, in this case, it would have quickly exposed the issue with `makeGalaxy`.

As far as I'm aware, though, there's no *requirement* in our procedures to rebase against master before the review; it's just something that I'm suggesting you should consider.

Comment (pgee2000) [inline tests/Kron.py]: I would guess in this case, given the complexity of the merge from HSC,
that it would have made sense.  This hasn't come up in the past, and as you
can see, I was completely unaware of it when I asked for review.

I think that I need to learn to at least look at the state of master before
sending something out for review, and that is a lesson learned.

Thanks.

Perry

Comment (PaulPrice) [inline tests/Kron.py]: Before you send something for review, you should review it yourself.

**Comment by jdswinbank** [comment]
This work is long since complete.
